File: src/illumination.jl
srcillum_helper(field_ginsu::Array{T}, time_mask::T, nthreads) where {T<:Float32} =
    ccall((:srcillum_helper_float,libillumination), Cvoid,
        (Ptr{Float32},Cfloat,Csize_t,Cint), field_ginsu, time_mask, length(field_ginsu), nthreads)

srcillum_helper(field_ginsu::Array{T}, time_mask::T, nthreads) where {T<:Float64} =
    ccall((:srcillum_helper_double,libillumination), Cvoid,
        (Ptr{Float64},Cdouble,Csize_t,Cint), field_ginsu, time_mask, length(field_ginsu), nthreads)

illum_accumulate(field_ginsu_accum::Array{T}, field_ginsu::Array{T}, time_mask::T, nthreads) where {T<:Float32} =
    ccall((:illum_accumulate_float,libillumination), Cvoid,
        (Ptr{Float32},Ptr{Float32},Cfloat,Csize_t,Cint), field_ginsu_accum, field_ginsu, time_mask, length(field_ginsu), nthreads)

illum_accumulate(field_ginsu_accum::Array{T}, field_ginsu::Array{T}, time_mask::T, nthreads) where {T<:Float64} =
    ccall((:illum_accumulate_double,libillumination), Cvoid,
        (Ptr{Float64},Ptr{Float64},Cdouble,Csize_t,Cint), field_ginsu_accum, field_ginsu, time_mask, length(field_ginsu), nthreads)

File: src/illumination.c
#include<stdio.h>

void
srcillum_helper_float(
    float* field_ginsu,
    float  time_mask,
    size_t n,
    int    nthreads)
{
    size_t i;
#pragma omp parallel for num_threads(nthreads)
    for (i = 0; i < n; i++) {
        field_ginsu[i] = time_mask*field_ginsu[i]*field_ginsu[i];
    }
}
void
srcillum_helper_double(
    double* field_ginsu,
    double  time_mask,
    size_t  n,
    int     nthreads)
{
    size_t i;
#pragma omp parallel for num_threads(nthreads)
    for (i = 0; i < n; i++) {
        field_ginsu[i] = time_mask*field_ginsu[i]*field_ginsu[i];
    }
}

void
illum_accumulate_float(
    float* field_ginsu_accum,
    float* field_ginsu,
    float  time_mask,
    size_t n,
    int    nthreads)
{
    size_t i;
#pragma omp parallel for num_threads(nthreads)
    for (i = 0; i < n; i++) {
        field_ginsu_accum[i] += time_mask*field_ginsu[i]*field_ginsu[i];
    }
}
void
illum_accumulate_double(
    double* field_ginsu_accum,
    double* field_ginsu,
    double  time_mask,
    size_t n,
    int    nthreads)
{
    size_t i;
#pragma omp parallel for num_threads(nthreads)
    for (i = 0; i < n; i++) {
        field_ginsu_accum[i] += time_mask*field_ginsu[i]*field_ginsu[i];
    }
}
